packages/firestore/test/unit/specs/listen_spec.test.ts

```diff
@@ -271,18 +271,54 @@ describeSpec('Listens:', [], () => {
         // (the document falls out) and send us a snapshot that's ahead of
         // docAv3 (which is already in our cache).
         .userListens(visibleQuery, 'resume-token-1000')
-        .watchAcksFull(visibleQuery, 5000, docAv2)
+        .watchAcks(visibleQuery)
+        .watchSends({ affects: [visibleQuery] }, docAv2)
+        .watchSnapshots(5000)
+        .watchSends({ affects: [visibleQuery] }, docAv3)
+        .watchCurrents(visibleQuery, 'resume-token-5000')
+        .watchSnapshots(6000)
         .expectEvents(visibleQuery, { fromCache: false })
         .userUnlistens(visibleQuery)
         .watchRemoves(visibleQuery)
         // Listen to allQuery again and make sure we still get docAv3.
         .userListens(allQuery, 'resume-token-4000')
         .expectEvents(allQuery, { added: [docAv3], fromCache: true })
-        .watchAcksFull(allQuery, 6000)
+        .watchAcksFull(allQuery, 7000)
         .expectEvents(allQuery, { fromCache: false })
     );
   });
 
+  specTest('Deleted documents in cache are fixed', [], () => {
+    const allQuery = Query.atPath(path('collection'));
+    const docAv1 = doc('collection/a', 1000, { key: 'a' });
+    const docDeleted = deletedDoc('collection/a', 2000);
+
+    return (
+      spec()
+        // Presuppose an initial state where the remote document cache has a
+        // broken synthesized delete at a timestamp later than the true version
+        // of the document. This requires both adding and later removing the
+        // document in order to force the watch change aggregator to propagate
+        // the deletion.
+        .withGCEnabled(false)
+        .userListens(allQuery)
+        .watchAcksFull(allQuery, 1000, docAv1)
+        .expectEvents(allQuery, { added: [docAv1], fromCache: false })
+        .watchSends({ removed: [allQuery] }, docDeleted)
+        .watchSnapshots(2000, [allQuery], 'resume-token-2000')
+        .watchSnapshots(2000)
+        .expectEvents(allQuery, { removed: [docAv1], fromCache: false })
+        .userUnlistens(allQuery)
+        .watchRemoves(allQuery)
+
+        // Now when the client listens expect the cached NoDocument to be
+        // discarded because the global snapshot version exceeds what came before.
+        .userListens(allQuery, 'resume-token-2000')
+        .watchAcksFull(allQuery, 3000, docAv1)
```

Comment:
Sorry - wouldn't this send an existence filter with count of 1 and not send this document (since the resume token is newer than the document)?

It looks like a query with a single NoDocument associated has an document count of 1, and we will not reset the query (only verified by glancing through the code).

Reply:
I've changed this to use a separate setup query to sidestep the existence filter issue.